File: packages/i18nify-go/modules/currency/format_compact_number.go
package currency

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

const (
	crore float64 = 10_000_000 // 1,00,00,000
	lakh  float64 = 100_000    // 1,00,000
)

// FormatCompactNumber formats amount using the Indian compact notation.
// Values ≥ 1 Crore are expressed as "X Cr"; values ≥ 1 Lakh as "X L".
// Values below 1 Lakh fall through to FormatNumber with the supplied opts.
// Mirrors the JS formatCompactNumber in i18nify-js/src/modules/currency/formatCompactNumber.ts.
func FormatCompactNumber(amount interface{}, opts NumberFormatOptions) (string, error) {
	val, err := ValidateAndConvertAmount(amount)
	if err != nil {
		return "", fmt.Errorf("parameter 'amount' is not a valid number: %v", err)
	}

	sign := ""
	abs := val
	if val < 0 {
		sign = "-"
		abs = -val
	}

	if abs >= crore {
		return sign + compactStr(abs/crore) + " Cr", nil
	}
	if abs >= lakh {
		return sign + compactStr(abs/lakh) + " L", nil
	}

	return FormatNumber(amount, opts)
}

// compactStr formats a compact coefficient, trimming insignificant trailing zeros.
// 1.00 → "1", 1.50 → "1.5", 1.53 → "1.53".
func compactStr(v float64) string {
	if v == math.Trunc(v) {
		return strconv.FormatInt(int64(v), 10)
	}
	s := strconv.FormatFloat(v, 'f', 2, 64)
	s = strings.TrimRight(s, "0")
	s = strings.TrimRight(s, ".")
	return s
}

File: packages/i18nify-js/src/modules/currency/formatCompactNumber.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { CurrencyCodeType, I18nifyNumberFormatOptions } from './types';
import formatNumber from './formatNumber';

const CRORE = 10_000_000; // 1,00,00,000
const LAKH = 100_000; // 1,00,000

const trimTrailingZeros = (n: number): string =>
  n.toFixed(2).replace(/\.?0+$/, '');

const formatCompactNumber = (
  amount: string | number,
  options: {
    currency?: CurrencyCodeType;
    locale?: string;
    intlOptions?: I18nifyNumberFormatOptions;
  } = {},
): string => {
  if (!Number(amount) && Number(amount) !== 0)
    throw new Error(
      `Parameter 'amount' is not a valid number. The received value was: ${amount} of type ${typeof amount}. Please ensure you pass a valid number.`,
    );

  const num = Number(amount);
  const sign = num < 0 ? '-' : '';
  const abs = Math.abs(num);

  if (abs >= CRORE) return `${sign}${trimTrailingZeros(abs / CRORE)} Cr`;
  if (abs >= LAKH) return `${sign}${trimTrailingZeros(abs / LAKH)} L`;

  return formatNumber(num, options);
};

export default withErrorBoundary<typeof formatCompactNumber>(
  formatCompactNumber,
);
